crates/nu-protocol/src/config/helper.rs

```diff
@@ -1,132 +1,169 @@
+        }
+    }
+}
+
+impl UpdateFromValue for usize {
+    fn update(&mut self, value: &Value, path: &mut ConfigPath, errors: &mut ConfigErrors) {
+        if let Ok(val) = value.as_int() {
+            if let Ok(val) = val.try_into() {
+                *self = val;
+            } else {
+                errors.invalid_value(path, "a non-negative integer", value);
             }
+        } else {
+            errors.type_mismatch(path, Type::Int, value);
         }
-    } else {
-        errors.push(ShellError::GenericError {
-            error: "Error while applying config changes".into(),
-            msg: format!("$env.config.{} should be a string", config_path.join(".")),
-            span: Some(span),
-            help: Some("This value will be ignored.".into()),
-            inner: vec![],
-        });
-        // Reconstruct
-        *value = config_point.clone().into_value(span);
     }
 }
 
-pub(super) fn process_bool_config(
-    value: &mut Value,
-    errors: &mut Vec<ShellError>,
-    config_point: &mut bool,
-) {
-    if let Ok(b) = value.as_bool() {
-        *config_point = b;
-    } else {
-        errors.push(ShellError::GenericError {
-            error: "Error while applying config changes".into(),
-            msg: "should be a bool".to_string(),
-            span: Some(value.span()),
-            help: Some("This value will be ignored.".into()),
-            inner: vec![],
-        });
-        // Reconstruct
-        *value = Value::bool(*config_point, value.span());
+impl UpdateFromValue for String {
+    fn update(&mut self, value: &Value, path: &mut ConfigPath, errors: &mut ConfigErrors) {
+        if let Ok(val) = value.as_str() {
+            *self = val.into();
+        } else {
+            errors.type_mismatch(path, Type::String, value);
+        }
     }
 }
 
-pub(super) fn process_int_config(
-    value: &mut Value,
-    errors: &mut Vec<ShellError>,
-    config_point: &mut i64,
-) {
-    if let Ok(b) = value.as_int() {
-        *config_point = b;
-    } else {
-        errors.push(ShellError::GenericError {
-            error: "Error while applying config changes".into(),
-            msg: "should be an int".into(),
-            span: Some(value.span()),
-            help: Some("This value will be ignored.".into()),
-            inner: vec![],
-        });
-        // Reconstruct
-        *value = Value::int(*config_point, value.span());
+impl<K, V> UpdateFromValue for HashMap<K, V>
+where
+    K: Borrow<str> + for<'a> From<&'a str> + Eq + Hash,
+    V: Default + UpdateFromValue,
+{
+    fn update<'a>(
+        &mut self,
+        value: &'a Value,
+        path: &mut ConfigPath<'a>,
+        errors: &mut ConfigErrors,
+    ) {
+        if let Ok(record) = value.as_record() {
+            self.retain(|k, _| record.contains(k.borrow()));
+            for (key, val) in record {
+                self.entry(key.as_str().into())
+                    .or_default()
+                    .update(val, path, errors);
```

Comment:
Is this more efficient than overwriting as a whole?

Reply:
Could you explain in more detail? Not 100% sure what you mean.

Reply:
Walking the whole record for each hashmap entry in `retain` and then walking the record to write the hashmap. This reuses the hashmap allocation but does more than simply collecting a fresh hashmap from the record once,

But also not sure if it is worth benchmarking :D

Reply:
Oh ok, I see. I think what I have now is better. I also fixed the wrong `path` being passed down.